### PR TITLE
Towards a complete example on MNIST with CNN

### PR DESCRIPTION
**Pull request by LaRiffle**
## Real world example of PySyft
The goal of this PR is to have a canonical example of PySyft in a "real world" setting, here MNIST + CNN.
We want to show how little changes there are to move from local pytorch to federated learning with pysyft.

## Roadmap to follow up on this PR

 -  **momentum args in optim SGD** #1896
 -  **use `FederatedDataset` & `FederatedDataLoader` @jonlebensold**
 -  Support .to(device) hooking #1893
 -  Warning in log_softmax -> no hook_kwargs

### CODE REVIEW

**Comment by iamtrask** [comment]
For this demo - can we move the MNIST pre-processing code to the sandbox?

**Comment by jlebensold** [comment]
@LaRiffle and I are working through an example.

I think that any MNIST demo needs to have the following design requirements:

1. The orchestrator (federated server) should not know about the individual datasets or tensors on each worker
2. CUDA support
3. An approach to messaging that is protocol independent and supports different implementations of the same code

**Comment by LaRiffle** [comment]
@jlebensold I believe that this MNIST demo is intended for people joining OM with a PyTorch background, so we should be as close as possible from the PyTorch official example to that people understand that FL can be turned on with very few changes.
 So I would rather focus on having a search + FederatedDataset + FederatedLoader stack for the data (very close to the official example)
and also I agree cuda would be cool, but I don't know how hard this is.

**Comment by iamtrask** [comment]
@LaRiffle - ok good point - maybe we can do a followon later which really leans on search as sortof "what deep learning will look like in the future." This would actually make for a killer blogpost too.

File: test/torch/test_federated_learning.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 hook = sy.TorchHook(torch)
 from torch import nn
-from syft import optim
+from torch import optim
```
Comment (iamtrask) [inline test/torch/test_federated_learning.py]: i can't believe you got this working! this is awesome!

**Pull request merged**